File: problems/Java/39. Combination Sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        bt(candidates, target, new ArrayList<>(), 0, res);
        return res;
    }

    private void bt(int[] candidates, int rem, List<Integer> path, int start, List<List<Integer>> res) {
        if (rem == 0) {
            res.add(new ArrayList<>(path));
            return;
        }
        if (rem < 0) {
            return;
        }

        for (int i = start; i < candidates.length; i++) {
            path.add(candidates[i]);
            bt(candidates, rem - candidates[i], path, i, res);
            path.remove(path.size() - 1);
        }
    }
}

File: problems/C++/39. Combination Sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> res;
        vector<int> path;
        bt(candidates, target, path, 0, res);
        return res;
    }

private:
    void bt(vector<int>& candidates, int rem, vector<int>& path, int start, vector<vector<int>>& res) {
        if (rem == 0) {
            res.push_back(path);
            return;
        }
        if (rem < 0) {
            return;
        }

        for (int i = start; i < candidates.size(); ++i) {
            path.push_back(candidates[i]);
            bt(candidates, rem - candidates[i], path, i, res);
            path.pop_back();
        }
    }
};